### PR TITLE
Feature/add resource manager

### PR DESCRIPTION
**Pull request by mlohvynenko**

### CODE REVIEW

File: include/aos/common/config.hpp [hunk 1]
```diff
@@ -127,6 +127,146 @@
 #define AOS_CONFIG_TYPES_FILE_CHUNK_SIZE 1024
 #endif
 
+/**
+ * File system mount destination len.
+ */
+#ifndef AOS_CONFIG_TYPES_FS_MOUNT_DESTINATION_LEN
+#define AOS_CONFIG_TYPES_FS_MOUNT_DESTINATION_LEN 48
```
Comment (mykola-kobets-epam) [inline include/aos/common/config.hpp:134]: Probably we should be more general here, please provide(if none) smth like cFSPathLen & utilize it for 
AOS_CONFIG_TYPES_FS_MOUNT_SOURCE_LEN / DESTINATION

File: include/aos/common/config.hpp [hunk 2]
```diff
@@ -127,6 +127,146 @@
...
+ * Host name len.
+ */
+#ifndef AOS_CONFIG_TYPES_HOST_NAME_LEN
+#define AOS_CONFIG_TYPES_HOST_NAME_LEN 48
+#endif
+
+/**
+ * Host device len.
+ */
+#ifndef AOS_CONFIG_TYPES_DEVICE_LEN
```
Comment (mykola-kobets-epam) [inline include/aos/common/config.hpp:182]: I would suggest device name len

File: include/aos/common/config.hpp [hunk 3]
```diff
@@ -127,6 +127,146 @@
...
+ * Max number of devices.
+ */
+#ifndef AOS_CONFIG_TYPES_MAX_NUM_DEVICES
+#define AOS_CONFIG_TYPES_MAX_NUM_DEVICES 8
+#endif
+
+/**
+ * Max number of resources.
+ */
+#ifndef AOS_CONFIG_TYPES_MAX_NUM_RESOURCES
```
Comment (mykola-kobets-epam) [inline include/aos/common/config.hpp:252]: resouce is a too abstract name, please be more specific

File: include/aos/common/config.hpp [hunk 4]
```diff
@@ -127,6 +127,146 @@
...
+#define AOS_CONFIG_TYPES_MAX_NUM_RESOURCES 4
+#endif
+
+/**
+ * Label len.
+ */
+#ifndef AOS_CONFIG_TYPES_LABEL_LEN
```
Comment (mykola-kobets-epam) [inline include/aos/common/config.hpp:259]: please be more specific about label name.

File: include/aos/common/config.hpp [hunk 5]
```diff
@@ -127,6 +127,146 @@
...
+#define AOS_CONFIG_TYPES_LABEL_LEN 16
+#endif
+
+/**
+ * Max number of labels.
+ */
+#ifndef AOS_CONFIG_TYPES_MAX_NUM_LABELS
```
Comment (mykola-kobets-epam) [inline include/aos/common/config.hpp:266]: same as above

File: src/sm/resourcemanager/log.hpp [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LOG_HPP_
+#define LOG_HPP_
+
+#include "aos/common/tools/log.hpp"
+
+#define LOG_DBG() LOG_MODULE_DBG(LogModuleEnum::eSMResourceManager)
```
Comment (mykola-kobets-epam) [inline src/sm/resourcemanager/log.hpp:13]: That's should be clarified whether we need a separate eSMResourceManager in logs or a single LogModuleEnum::eSM per sm component would be enough.

Comment (mlohvynenko) [reply src/sm/resourcemanager/log.hpp:13]: aos core zephyr has a separate enum value for the resource manager
https://github.com/aoscloud/aos_core_zephyr/blob/main/src/resourcemanager/log.hpp#L13
`Logger::Module::eResourceMgr`

That's why I've introduced a new one in the core lib cpp.

@al1img , can you please suggest on the approach? Should  I keep using the new enum values, or re-use LogModuleEnum::eSM?

File: include/aos/sm/resourcemanager.hpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+#include <aos/common/tools/string.hpp>
+
+/**
+ * Resource manager interface.
+ */
+
+class ResourceManagerItf {
+public:
+    /**
+     * Destructor
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply include/aos/sm/resourcemanager.hpp:138]: This commit contains aos_core_zephyr implementation + some minor changes to make it compilable.
I'd like to keep the code in this commit unchanged.
All these suggestions are applied in the next commits.

File: include/aos/sm/resourcemanager.hpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    virtual ~ResourceManagerItf() = default;
+
+    /**
+     * Get current unit config version.
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply include/aos/sm/resourcemanager.hpp]: This commit contains aos_core_zephyr implementation + some minor changes to make it compilable.
I'd like to keep the code in this commit unchanged.
All these suggestions are applied in the next commits.

File: include/aos/sm/resourcemanager.hpp [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+     *
+     * @param[out] version param to store version.
+     * @return aos::Error.
+     */
+    virtual aos::Error GetUnitConfigInfo(aos::String& version) const = 0;
+
+    /**
+     * Check new unit configuration.
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply include/aos/sm/resourcemanager.hpp]: This commit contains aos_core_zephyr implementation + some minor changes to make it compilable.
I'd like to keep the code in this commit unchanged.
All these suggestions are applied in the next commits.

File: include/aos/sm/resourcemanager.hpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+     *
+     * @param version unit config version
+     * @param unitConfig string with unit configuration.
+     * @return aos::Error.
+     */
+    virtual aos::Error CheckUnitConfig(const aos::String& version, const aos::String& unitConfig) const = 0;
+
+    /**
+     * Update unit configuration.
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply include/aos/sm/resourcemanager.hpp]: This commit contains aos_core_zephyr implementation + some minor changes to make it compilable.
I'd like to keep the code in this commit unchanged.
All these suggestions are applied in the next commits.

File: include/aos/sm/resourcemanager.hpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+/**
+ * Resource manager instance.
+ */
+
+class ResourceManager : public ResourceManagerItf, private aos::NonCopyable {
+public:
+    /**
+     * Get current unit config version.
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply include/aos/sm/resourcemanager.hpp]: This commit contains aos_core_zephyr implementation + some minor changes to make it compilable.
I'd like to keep the code in this commit unchanged.
All these suggestions are applied in the next commits.

**Review by MykolaSolyanko — APPROVED** [review]

File: include/aos/common/config.hpp [hunk 12]
```diff
@@ -625,4 +625,11 @@
 #define AOS_CONFIG_CRYPTOPROVIDER_PUB_KEYS_COUNT 5
 #endif
 
+/**
+ * Unit config JSON length.
+ */
+#ifndef AOS_CONFIG_UNIT_CONFIG_JSON_LEN
+#define AOS_CONFIG_UNIT_CONFIG_JSON_LEN 256
```
Comment (al1img) [inline include/aos/common/config.hpp:632]: Better to set something like 4096

**Review by mykola-kobets-epam — APPROVED** [review]

File: include/aos/sm/resourcemanager.hpp [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    virtual ~ResourceManagerItf() = default;
+
+    /**
+     * Get current unit config version.
+     *
+     * @param[out] version param to store version.
+     * @return aos::Error.
+     */
+    virtual aos::Error GetUnitConfigInfo(aos::String& version) const = 0;
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply include/aos/sm/resourcemanager.hpp]: Applied suggestioon

**Pull request merged**